transport: remove decodeState from client to reduce allocations

JNProtzman opened the pull request:
Remove decodeState from http_client, and decode the header fields within the client. 

This will fix half of #1457. I can submit a second PR to remove decodeState from http_server, or if it's preferred I can fold it into this PR too.

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1179,14 +1181,139 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+		encoding       string
+		contentTypeErr string
+		rawStatusMsg   string
+	)
```
Is it possible to eliminate some/all/most of these vars by doing the logic in-line?

E.g. instead of setting `isGRPC` and then checking it after the loop, move the contents of the `if isGRPC` to where we set it.

JNProtzman replied:
I was able to remove the grpcErr and httpErr fields. I've tried to remove other fields but wasn't able to. Please let me know what you think!

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1179,14 +1181,153 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+		case "grpc-status":
+			code, err := strconv.Atoi(hf.Value)
+			if err != nil {
+				if isGRPC {
```
This is ordering dependent (content-type must come first).  It's an existing assumption in the code that has apparently held true for us, so it is fine to not fix this in this PR.

I have an idea to help remove some of the temporary variables.  I'm not certain whether it will be better, but maybe give it a shot: build the `mdata` map including literally everything, then do some post-processing afterwards, deleting the special-cased things where we'd skip adding it to the map in this loop.

So:

```go
for _, hf := range frame.Fields {
  // decode, check reserved/whitelist, etc
  mdata[hf.Name] = append(....)
}
if ct := mdata["content-type"]; len(ct) != 1 {
  // error
} else {
  if _, validCT := contentSubtype(ct[0]); !validCT {
    // error
  }
}

if s, ok := mdata["grpc-status"]; !ok || len(s) != 1 {
   // error
} else {
  // use s and check grpc-message
}
delete(mdata, "grpc-status")
delete(mdata, "grpc-message") // etc..or do this with a for loop, only allowing
                              // non-reserved or whitelisted things to remain.
```

This could help remove some special-case things like the errors decoding binary headers (grpc-tags-bin, grpc-trace-bin, grpc-status-details-bin), since those would fail naturally when processing all headers at the top.

dfawley replied:
Ping @JNProtzman - any thoughts on this / updates?

JNProtzman replied:
Hey Doug, sorry for the delay. I worked on this a bit over the long weekend. I have a couple follow up questions:
1) What would the error be if "content-type" or "grpc-status" aren't passed in as parameters? There is nothing that currently checks that / fails if those headers aren't passed in. 

I can attempt to remove the variables by using the map like you've suggested. I'll try to have a commit against this PR for it this week.

JNProtzman commented:
I've removed most of the local variables. There is one left that I can remove, but not sure about doing the work twice - also not sure about the efficiency of running strconv twice?
I need to leave isGRPC in so that we can error appropriately in the for loops, otherwise I could add everything to the map, then process it after? Let me know what you think!

dfawley commented:
@JNProtzman I see a change was committed since my review but it appears to be just a merge of master into this branch.  Have you had a chance to update since the latest review comments?  Thanks!

JNProtzman commented:
I've reduced the amount of code in the loop, as well as run the benchmark:
```                                                                                 
unary-networkMode_Local-bufConn_false-keepalive_false-benchTime_30s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_100-reqSize_1B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps      1445138      1616990    11.89%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op      8499.22      8934.02     5.12%
           Allocs/op       159.75       162.83     1.88%
             ReqT/op    385370.13    431197.33    11.89%
            RespT/op    385370.13    431197.33    11.89%
            50th-Lat   1.816484ms    1.75466ms    -3.40%
            90th-Lat    3.16187ms   2.776928ms   -12.17%
            99th-Lat   5.520533ms   4.004501ms   -27.46%
             Avg-Lat   2.074931ms   1.854524ms   -10.62%
           GoVersion     go1.14.1     go1.14.1
         GrpcVersion   1.29.0-dev   1.29.0-dev
```

JNProtzman commented on internal/transport/http2_client.go:
```diff
@@ -1180,14 +1182,112 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+	// list size hits MaxHeaderListSize limit.
+	if frame.Truncated {
+		err := status.Error(codes.Internal, "peer header list size exceeded limit")
 		t.closeStream(s, err, true, http2.ErrCodeProtocol, status.Convert(err), nil, endStream)
```
Done!

JNProtzman commented on internal/transport/http2_client.go:
```diff
@@ -1180,14 +1182,112 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+
+	for _, hf := range frame.Fields {
+		switch hf.Name {
+		case "content-type":
```
I've reduced the loop as you suggested.

JNProtzman commented:
After creating the statuses, here is the benchmark:
```
unary-networkMode_Local-bufConn_false-keepalive_false-benchTime_30s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_100-reqSize_1B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps      1445138      1612105    11.55%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op      8499.22      8934.46     5.12%
           Allocs/op       159.75       162.86     1.88%
             ReqT/op    385370.13    429894.67    11.55%
            RespT/op    385370.13    429894.67    11.55%
            50th-Lat   1.816484ms   1.766895ms    -2.73%
            90th-Lat    3.16187ms   2.777164ms   -12.17%
            99th-Lat   5.520533ms    3.93908ms   -28.65%
             Avg-Lat   2.074931ms   1.859758ms   -10.37%
           GoVersion     go1.14.1     go1.14.1
         GrpcVersion   1.29.0-dev   1.29.0-dev
```

JNProtzman commented:
Friendly ping @dfawley

JNProtzman commented:
Ping @dfawley

dfawley commented:
Sorry for the delay..  I still have several things on my plate before I'll be able to get to this.

JNProtzman commented:
@dfawley - just checking in since it's been a few months.
If you don't have time, no worries - I'll do my best to keep this updated / conflict free for when you have some spare cycles.

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1203,14 +1205,93 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+			if logger.V(logLevel) {
+				logger.Errorf("Failed to decode metadata header (%q, %q): %v", hf.Name, hf.Value, err)
+			}
+			mdata[hf.Name] = []string{}
+			break
+		}
+		mdata[hf.Name] = append(mdata[hf.Name], v)
```
I'm surprised the performance is actually better with this change.  Using a `map` instead of a `struct` seems like it would be worse.

Any ideas?

Can we reduce the map accesses and make it even better?
```go
var contentType string
var grpcMessage string
for _, hf := range frame.Fields {
	....
	switch hf.Name {
	case "content-type":
		if contentType != "" {
			fail?
		}
		contentType = v
	case "grpc-message":
		if grpcMessage != "" {
			fail?
		}
		grpcMessage = v
	....
	default:
		mdata[hf.Name] = append(..., v)
	}
}
```

JNProtzman replied:
I feel like my most recent push is a regression. I know we had gotten rid of some fields in favor of map-accesses but in order to reduce / remove the map accesses, we can create the necessary local fields. This has the added benefit of being a pretty clean solution (it's basically just removing the decodeState abstraction).

I did run the benchmarks again, this time for a minute instead of 30s. Here's those results:

```
unary-networkMode_Local-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_100-reqSize_1B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps      3074579      3197715     4.00%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op      8854.32      8892.72     0.43%
           Allocs/op       163.45       163.48     0.00%
             ReqT/op    409943.87    426362.00     4.01%
            RespT/op    409943.87    426362.00     4.01%
            50th-Lat   1.845107ms     1.8044ms    -2.21%
            90th-Lat   2.852592ms   2.674804ms    -6.23%
            99th-Lat   4.551656ms   3.953715ms   -13.14%
             Avg-Lat   1.949743ms   1.874708ms    -3.85%
           GoVersion     go1.16.2     go1.16.2
         GrpcVersion   1.38.0-dev   1.38.0-dev
```

dfawley replied:
This is certainly not as good as last time.  I wonder if the previous run was a fluke, or maybe this time was a fluke?  There can be a lot of variance in the benchmarks.  I'll take a look at the code again and see if I can spot anything.

dfawley commented on internal/transport/http_util.go:
```diff
@@ -293,6 +285,26 @@ func (d *decodeState) constructHTTPErrMsg() string {
 
+// constructErrMsg constructs error message to be returned in HTTP fallback mode.
+// Format: HTTP status code and its corresponding message + content-type error message.
+func constructHTTPErrMsg(httpStatus *int, contentTypeErr string) string {
```
This looks identical to the above; if so, please make `decodeState.constructHTTPErrMsg` call this one instead of copy/pasting.

JNProtzman replied:
Done!

dfawley reviewed (changes requested):
Sorry for the absurdly long delay here.  This was lost in the fray, and we are having a fixit this week to catch up.  This looks mostly good, if you are still interested in pursuing, I think we can wrap it up quickly.  Thanks!

JNProtzman commented:
I'm open to thinking about this a little bit longer to try to get the best solution. It's not the most simple code and the nice thing about the map is that we can get the fields - specifically `content-subtype` - in the order we want to get the code to be more linear.

Please let me know what you think the next steps should be.

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1254,14 +1256,132 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+				code = codes.Unknown
+			}
+		}
+		if err := status.Error(code, constructHTTPErrMsg(
+			httpStatus,
+			contentTypeErr,
+		)); err != nil {
```
Why is this an `if != nil` if you are constructing it?  It can be simply:

```go
err := status.Error(code, construct(...))
t.closeStream(s, err, ...)
return
```

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1254,14 +1256,132 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+		var code = codes.Internal // when header does not include HTTP status, return INTERNAL
+		if httpStatus != nil {
+			var ok bool
+			code, ok = HTTPStatusConvTab[*(httpStatus)]
```
Parens should not be necessary here.

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1254,14 +1256,132 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+		// Unknown status will be propagated to the user.
+		// Otherwise, it will be ignored. In which case, status from
+		// a later trailer, that has StreamEnded flag set, is propagated.
+		if rawStatusCode == nil && statusGen == nil {
+			code := int(codes.Unknown)
+			rawStatusCode = &code
```
How about:

```
rawStatusCode := codes.Unknown
....
case "grpc-status":
	code, err := strconv.ParseUint(hf.Value, 0, 32)
	if err != nil {
		grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-status: %v", err))
		break
	}
	rawStatusCode = codes.Code(uint32(code))
...
```

and remove this code here?  This whole `if isGRPC` part can be removed, leaving the `else` (`if !isGRPC`) by itself.

JNProtzman commented:
I've addressed the PR comments - at this point the benchmarks show a very similar performance gain compared to the update yesterday:

```unary-networkMode_Local-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_100-reqSize_1B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps      2832020      2946728     4.05%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op      8858.60      8889.01     0.35%
           Allocs/op       163.41       162.46    -0.61%
             ReqT/op    377602.67    392897.07     4.05%
            RespT/op    377602.67    392897.07     4.05%
            50th-Lat   1.875847ms    1.84894ms    -1.43%
            90th-Lat   3.127691ms   3.001616ms    -4.03%
            99th-Lat   6.367377ms    5.66712ms   -11.00%
             Avg-Lat    2.11629ms   2.034435ms    -3.87%
           GoVersion     go1.16.2     go1.16.2
         GrpcVersion   1.38.0-dev   1.38.0-dev
```

I'll play around a bit and look at the code at the commit where the performance improvement was double what I'm seeing now.

JNProtzman commented:
OK one more change - I moved the ":status" header processing out of the switch which removes one Atoi conversion. I ran the benchmark twice and here is the second, slower benchmark:
```
unary-networkMode_Local-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_100-reqSize_1B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps      2832020      3053091     7.81%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op      8858.60      8895.10     0.42%
           Allocs/op       163.41       162.47    -0.61%
             ReqT/op    377602.67    407078.80     7.81%
            RespT/op    377602.67    407078.80     7.81%
            50th-Lat   1.875847ms   1.828831ms    -2.51%
            90th-Lat   3.127691ms   2.882664ms    -7.83%
            99th-Lat   6.367377ms   4.990425ms   -21.63%
             Avg-Lat    2.11629ms    1.96374ms    -7.21%
           GoVersion     go1.16.2     go1.16.2
         GrpcVersion   1.38.0-dev   1.38.0-dev

```

dfawley commented:
> here is the second, slower benchmark

Not sure what you mean by "slower"?  It seems much faster than the previous one.

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1254,14 +1256,122 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+		case "grpc-encoding":
+			s.recvCompress = hf.Value
+		case "grpc-status":
+			code, err := strconv.Atoi(hf.Value)
```
Please use `strconv.ParseInt(..., 32)` to avoid the code scanning warning below.

JNProtzman replied:
Done

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1254,14 +1256,122 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+				contentTypeErr = fmt.Sprintf("transport: received the unexpected content-type %q", hf.Value)
+				break
+			}
+			// TODO: do we want to propagate the whole content-type in the metadata,
+			// or come up with a way to just propagate the content-subtype if it was set?
+			// ie {"content-type": "application/grpc+proto"} or {"content-subtype": "proto"}
+			// in the metadata?
```
Delete this TODO, please. I think what we're doing is fine.

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1254,14 +1256,122 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+			}
+			statusGen = status.FromProto(st)
+		case ":status":
+			// we only process :status if this is not grpc so no need to ALWAYS convert the value
```
Don't we need to validate that there is a `:status` of 200 in gRPC mode?  Were we not doing this before?  If so, seems like a bug.

https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses

JNProtzman replied:
[On master](https://github.com/grpc/grpc-go/blob/master/internal/transport/http_util.go#L248), it looks like when we're in gRPC mode, the `:status` header is not evaluated.

The raw status code (`grpc-status` header) is evaluated - I'm not sure if that's a replacement, however.

JNProtzman replied:
Should I add a TODO or create an issue for it?

JNProtzman replied:
This issue will be fixed by  #4474. Resolving this comment.

JNProtzman commented:
Sorry for the confusion, I 

> > here is the second, slower benchmark
> 
> Not sure what you mean by "slower"? It seems much faster than the previous one.

Sorry for the confusion, I ran two benchmarks after moving the “:status” header processing and that was the slower of the two.

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1254,14 +1256,122 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+			statusGen = status.FromProto(st)
+		case ":status":
+			// we only process :status if this is not grpc so no need to ALWAYS convert the value
+			mdata[hf.Name] = append(mdata[hf.Name], hf.Value)
```
I would think using a local variable for this would still be better than adding/deleting in the map.

JNProtzman replied:
Reverted to a local variable

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1254,14 +1256,122 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+		case ":status":
+			// we only process :status if this is not grpc so no need to ALWAYS convert the value
+			mdata[hf.Name] = append(mdata[hf.Name], hf.Value)
+		case "grpc-tags-bin":
+			v, err := decodeBinHeader(hf.Value)
+			if err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-tags-bin: %v", err))
+				break
+			}
+			mdata[hf.Name] = append(mdata[hf.Name], string(v))
+		case "grpc-trace-bin":
+			v, err := decodeBinHeader(hf.Value)
+			if err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-trace-bin: %v", err))
+				break
+			}
+			mdata[hf.Name] = append(mdata[hf.Name], string(v))
```
These two are essentially the same.  Please combine them (`case "grpc-tags-bin", "grpc-trace-bin":`)

dfawley replied:
Actually, I _think_ these can be handled by the `default` case, too, since `decodeMetadataHeader` handles `-bin` suffixes.  Let's do that then if it works.

JNProtzman replied:
These can be handled by the default case, and it definitely cleans up the code a bit!

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1254,14 +1256,122 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+		case "grpc-message":
+			grpcMessage = decodeGrpcMessage(hf.Value)
+		case "grpc-status-details-bin":
+			v, err := decodeBinHeader(hf.Value)
+			if err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-status-details-bin: %v", err))
+				break
+			}
+			st := &spb.Status{}
+			if err := proto.Unmarshal(v, st); err != nil {
+				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-status-details-bin: %v", err))
+				break
+			}
+			statusGen = status.FromProto(st)
```
Maybe like the above, we can have `grpcStatus = decodeGRPCStatusDetails(hf.Value)` to clean up the code a bit?

JNProtzman replied:
done!

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1254,14 +1256,122 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+			v, err := decodeMetadataHeader(hf.Name, hf.Value)
+			if err != nil {
+				if logger.V(logLevel) {
+					logger.Errorf("Failed to decode metadata header (%q, %q): %v", hf.Name, hf.Value, err)
```
`Error` logs shouldn't be behind >0 verbosity levels.  Let's promote this to `Warning` always since it potentially means lost data for the application, but was hidden before and I'm worried it _could_ be spammy, so `Error` with no verbosity might be too much.

JNProtzman replied:
Done

JNProtzman commented:
These benchmarks are all over the place. Here's the latest after the recent set of changes
```
unary-networkMode_Local-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_100-reqSize_1B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps      2832020      3237437    14.32%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op      8858.60      8877.16     0.21%
           Allocs/op       163.41       161.55    -1.22%
             ReqT/op    377602.67    431658.27    14.32%
            RespT/op    377602.67    431658.27    14.32%
            50th-Lat   1.875847ms   1.762341ms    -6.05%
            90th-Lat   3.127691ms   2.654554ms   -15.13%
            99th-Lat   6.367377ms    4.05728ms   -36.28%
             Avg-Lat    2.11629ms   1.851927ms   -12.49%
           GoVersion     go1.16.2     go1.16.2
         GrpcVersion   1.38.0-dev   1.38.0-dev
```

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1282,96 +1283,73 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
 			grpcMessage = decodeGrpcMessage(hf.Value)
 		case "grpc-status-details-bin":
-			v, err := decodeBinHeader(hf.Value)
+			sg, err := decodeGRPCStatusDetails(hf.Value)
```
It's a style nit, but we usually prefer to avoid the new local variable:
```go
var err error
statusGen, err = decodeGRPCStatusDetails(hf.Value)
...
```

Then you also don't need the `break`

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1265,14 +1263,17 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
 		statusGen      *status.Status
-		rawStatusCode  = codes.Unknown
+
+		grpcStatus string
```
Maybe rename to `headerError` or something, with a comment like `// set if an error is encountered parsing headers`?

JNProtzman replied:
Done

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1282,96 +1283,73 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+			se := status.New(codes.Internal, grpcStatus)
+			t.closeStream(s, se.Err(), true, http2.ErrCodeProtocol, se, nil, endStream)
 			return
 		}
```
Should this be moved outside the loop?

If so, then we can remove the `if isGRPC` part of it by putting it after the `if !isGRPC` block.

dfawley reviewed (commented):
A few last small things otherwise this LGTM and I'm surprised and excited by the huge gain in performance.

JNProtzman commented:
Potentially the final benchmark:
```
               Title       Before        After Percentage
            TotalOps      2832020      3102375     9.55%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op      8858.60      8878.35     0.23%
           Allocs/op       163.41       161.48    -1.22%
             ReqT/op    377602.67    413650.00     9.55%
            RespT/op    377602.67    413650.00     9.55%
            50th-Lat   1.875847ms   1.823108ms    -2.81%
            90th-Lat   3.127691ms   2.797689ms   -10.55%
            99th-Lat   6.367377ms   4.606214ms   -27.66%
             Avg-Lat    2.11629ms   1.932541ms    -8.68%
           GoVersion     go1.16.2     go1.16.2
         GrpcVersion   1.38.0-dev   1.38.0-dev
```

JNProtzman commented on internal/transport/http2_client.go:
```diff
@@ -1254,11 +1254,97 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
+
+		// Verify the HTTP response is a 200.
+		se := status.Error(code, constructHTTPErrMsg(&httpStatusCode, contentTypeErr))
+		t.closeStream(s, se, true, http2.ErrCodeProtocol, status.Convert(se), nil, endStream)
```
Done!

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1337,8 +1337,8 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
-		se := status.Error(code, constructHTTPErrMsg(&httpStatusCode, contentTypeErr))
-		t.closeStream(s, se, true, http2.ErrCodeProtocol, status.Convert(se), nil, endStream)
+		se := status.New(code, constructHTTPErrMsg(&httpStatusCode, contentTypeErr))
+		t.closeStream(s, se.Err(), true, http2.ErrCodeProtocol, se, nil, endStream)
```
Oh, right, there is a type change happening.  This way is better, still, to avoid the extra copy.

dfawley reviewed (approved):
Thanks! 🎉

JNProtzman commented:
Once this is in, should I start working on the server side?

dfawley commented:
> Once this is in, should I start working on the server side?

Yes, that would be great, thanks.  Hopefully we get a similar speedup there.
